libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1697,9 +1697,43 @@ smtutil::Expression SMTEncoder::division(smtutil::Expression _left, smtutil::Exp
 {
 	// Signed division in SMTLIB2 rounds differently for negative division.
 	if (_type.isSigned())
-		return signedDivisionEVM(_left, _right);
+		return signedDivisionWithSlacks(_left, _right, _type); // instead of signedDivisionEVM(_left, _right);
 	else
-		return _left / _right;
+		return unsignedDivisionWithSlacks(_left, _right, _type); // instead of _left / _right;
+}
+
+smtutil::Expression SMTEncoder::unsignedDivisionWithSlacks(smtutil::Expression _left, smtutil::Expression _right, IntegerType const& _type)
+{
+	solAssert(!_type.isSigned(), "");
+	IntegerType const* intType = &_type;
+	string suffix = "div_" + to_string(m_context.newUniqueId());
+	smt::SymbolicIntVariable d(intType, intType, "d_" + suffix, m_context);
+	smt::SymbolicIntVariable r(intType, intType, "r_" + suffix, m_context);
+
+	// x / y = d iff d * y + r = x and 0 <= r < y
+	m_context.addAssertion(((d.currentValue() * _right) + r.currentValue()) == _left);
+	m_context.addAssertion(0 <= r.currentValue() && r.currentValue() < _right);
+	return d.currentValue();
```

Comment:
If ``_right`` (resp ``y``) is zero, the solver is now free to choose any value for ``d``...
Would it be more correct to change this to ``ite(_right == 0, 0, d)``? (EVM division by zero is defined to be zero)

Reply:
That and the second constraint is false. So yea, I think the `ite` makes sense

Reply:
True - ``r.currentValue() < _right`` should probably only be asserted conditionally, i.e.
``m_context.addAssertion(0 <= r.currentValue() && (_right == 0 || r.currentValue() < _right));`` or sth like that.

Reply:
Yes, if `_right` is zero than the second constraint is equivalent to false. So there will actually be **no** value of `d` the solver can choose to make this satisfiable.

Reply:
Yea, agree with conditionally adding the second constraint

Reply:
In my intended context of SMTEncoder::arithmeticOperation there is additional assertion passed to the solver that `_right` is not zero. But if this encoding would be used also elsewhere then I guess the `ite`  and conditional constraint make sense, I agree.

Reply:
Actually, you're right - and while EVM defines division by zero to be zero, solidity codegen actually reverts, so it's correct that the assertion is there in ``SMTEncoder::arithmeticOperation``.
So given that, you were probably right to begin with and none of this needs to be changed after all.

Reply:
Ah true. I think Chris wrote that code then, so you're right, I think it should stay.

Reply:
What we could do is to move the _right != 0 assertion here into this function, though. And document this in a comment in the header or sth like that, s.t. we hopefully won't start wondering about it again in the future :-) (then it'd need to move for the mod case as well, though).
But also doing nothing is fine with me, just a suggestion.

Reply:
It depends on the intended semantics of the encoding. If it should reflect the EVM semantics that I think the `ite` version would be best. I am not sure about the solidity semantics with the revert. I am now inclined to see it as a low-level method that provides the basic encoding and it is up to the caller to decide on additional constraints (this would be the current approach used in the code, I think).